In [ ]:
import pandas as pd
from scipy.stats import chi2_contingency
from statsmodels.stats.outliers_influence import variance_inflation_factor

In [16]:
a1 = pd.read_excel("C:/Github/Credit Risk Prediction/data/raw/case_study1.xlsx")
a2 = pd.read_excel("C:/Github/Credit Risk Prediction/data/raw/case_study2.xlsx")

In [17]:
df1 = a1.copy()
df2 = a2.copy()

In [18]:
df1.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 51336 entries, 0 to 51335
Data columns (total 26 columns):
 #   Column                Non-Null Count  Dtype  
---  ------                --------------  -----  
 0   PROSPECTID            51336 non-null  int64  
 1   Total_TL              51336 non-null  int64  
 2   Tot_Closed_TL         51336 non-null  int64  
 3   Tot_Active_TL         51336 non-null  int64  
 4   Total_TL_opened_L6M   51336 non-null  int64  
 5   Tot_TL_closed_L6M     51336 non-null  int64  
 6   pct_tl_open_L6M       51336 non-null  float64
 7   pct_tl_closed_L6M     51336 non-null  float64
 8   pct_active_tl         51336 non-null  float64
 9   pct_closed_tl         51336 non-null  float64
 10  Total_TL_opened_L12M  51336 non-null  int64  
 11  Tot_TL_closed_L12M    51336 non-null  int64  
 12  pct_tl_open_L12M      51336 non-null  float64
 13  pct_tl_closed_L12M    51336 non-null  float64
 14  Tot_Missed_Pmnt       51336 non-null  int64  
 15  Auto_TL            

df1 is data from CIBIL, there are no null values

In [19]:
df2.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 51336 entries, 0 to 51335
Data columns (total 62 columns):
 #   Column                        Non-Null Count  Dtype  
---  ------                        --------------  -----  
 0   PROSPECTID                    51336 non-null  int64  
 1   time_since_recent_payment     51336 non-null  int64  
 2   time_since_first_deliquency   51336 non-null  int64  
 3   time_since_recent_deliquency  51336 non-null  int64  
 4   num_times_delinquent          51336 non-null  int64  
 5   max_delinquency_level         51336 non-null  int64  
 6   max_recent_level_of_deliq     51336 non-null  int64  
 7   num_deliq_6mts                51336 non-null  int64  
 8   num_deliq_12mts               51336 non-null  int64  
 9   num_deliq_6_12mts             51336 non-null  int64  
 10  max_deliq_6mts                51336 non-null  int64  
 11  max_deliq_12mts               51336 non-null  int64  
 12  num_times_30p_dpd             51336 non-null  int64  
 13  n

df2 is dataset from bank records, here also there are no null values

In [20]:
df1.sample(5)

PROSPECTID  Total_TL  Tot_Closed_TL  Tot_Active_TL  \
1656         1657         4              0              4   
14425       14426         1              0              1   
17844       17845         2              1              1   
1498         1499         5              0              5   
30138       30139         1              0              1   

       Total_TL_opened_L6M  Tot_TL_closed_L6M  pct_tl_open_L6M  \
1656                     2                  0              0.5   
14425                    0                  0              0.0   
17844                    0                  0              0.0   
1498                     2                  0              0.4   
30138                    1                  0              1.0   

       pct_tl_closed_L6M  pct_active_tl  pct_closed_tl  ...  CC_TL  \
1656                 0.0            1.0            0.0  ...      0   
14425                0.0            1.0            0.0  ...      0   
17844                0.0            0.5            0.5  ...      0   
1498                 0.0            1.0            0.0  ...      0   
30138                0.0            1.0            0.0  ...      0   

       Consumer_TL  Gold_TL  Home_TL  PL_TL  Secured_TL  Unsecured_TL  \
1656             1        0        0      1           1             3   
14425            0        0        0      0           1             0   
17844            0        0        0      0           2             0   
1498             1        0        0      0           0             5   
30138            0        0        0      1           0             1   

       Other_TL  Age_Oldest_TL  Age_Newest_TL  
1656          1             19              4  
14425         0             15             15  
17844         1             55             10  
1498          4             17              4  
30138         0              6              6  

[5 rows x 26 columns]

nothing out of ordinary in this

In [21]:
df2.sample(5)

PROSPECTID  time_since_recent_payment  time_since_first_deliquency  \
44805       44806                        159                       -99999   
45393       45394                        405                       -99999   
21078       21079                        671                       -99999   
31862       31863                         84                           13   
16650       16651                         45                       -99999   

       time_since_recent_deliquency  num_times_delinquent  \
44805                        -99999                     0   
45393                        -99999                     0   
21078                        -99999                     0   
31862                            13                     1   
16650                        -99999                     0   

       max_delinquency_level  max_recent_level_of_deliq  num_deliq_6mts  \
44805                 -99999                          0               0   
45393                 -99999                          0               0   
21078                 -99999                          0               0   
31862                     26                         26               0   
16650                 -99999                          0               0   

       num_deliq_12mts  num_deliq_6_12mts  ...  pct_CC_enq_L6m_of_L12m  \
44805                0                  0  ...                     0.0   
45393                0                  0  ...                     0.0   
21078                0                  0  ...                     0.0   
31862                0                  0  ...                     0.0   
16650                0                  0  ...                     0.5   

       pct_PL_enq_L6m_of_ever  pct_CC_enq_L6m_of_ever  \
44805                     0.0                     0.0   
45393                     0.0                     0.0   
21078                     0.0                     0.0   
31862                     0.0                     0.0   
16650                     1.0                     0.5   

       max_unsec_exposure_inPct  HL_Flag  GL_Flag  last_prod_enq2  \
44805                -99999.000        0        0    ConsumerLoan   
45393                -99999.000        0        0    ConsumerLoan   
21078                -99999.000        0        0              AL   
31862                -99999.000        0        0          others   
16650                     2.919        0        0    ConsumerLoan   

       first_prod_enq2  Credit_Score  Approved_Flag  
44805     ConsumerLoan           681             P2  
45393           others           680             P2  
21078               AL           686             P2  
31862           others           672             P2  
16650     ConsumerLoan           641             P4  

[5 rows x 62 columns]

can see a lot of "-99999" values which are not null values, but just a placeholder for missing values.
rechecking if df1 also has some values as "-99999"

In [29]:
df1.describe()

PROSPECTID      Total_TL  Tot_Closed_TL  Tot_Active_TL  \
count  51336.000000  51336.000000   51336.000000   51336.000000   
mean   25668.500000      4.858598       2.770415       2.088184   
std    14819.571046      7.177116       5.941680       2.290774   
min        1.000000      1.000000       0.000000       0.000000   
25%    12834.750000      1.000000       0.000000       1.000000   
50%    25668.500000      2.000000       1.000000       1.000000   
75%    38502.250000      5.000000       3.000000       3.000000   
max    51336.000000    235.000000     216.000000      47.000000   

       Total_TL_opened_L6M  Tot_TL_closed_L6M  pct_tl_open_L6M  \
count         51336.000000       51336.000000     51336.000000   
mean              0.736754           0.428919         0.184574   
std               1.296717           0.989972         0.297414   
min               0.000000           0.000000         0.000000   
25%               0.000000           0.000000         0.000000   
50%               0.000000           0.000000         0.000000   
75%               1.000000           1.000000         0.308000   
max              27.000000          19.000000         1.000000   

       pct_tl_closed_L6M  pct_active_tl  pct_closed_tl  ...         CC_TL  \
count       51336.000000   51336.000000   51336.000000  ...  51336.000000   
mean            0.089095       0.577542       0.422458  ...      0.124981   
std             0.205635       0.379867       0.379867  ...      0.505201   
min             0.000000       0.000000       0.000000  ...      0.000000   
25%             0.000000       0.250000       0.000000  ...      0.000000   
50%             0.000000       0.556000       0.444000  ...      0.000000   
75%             0.053000       1.000000       0.750000  ...      0.000000   
max             1.000000       1.000000       1.000000  ...     27.000000   

        Consumer_TL       Gold_TL       Home_TL         PL_TL    Secured_TL  \
count  51336.000000  51336.000000  51336.000000  51336.000000  51336.000000   
mean       1.136084      1.561847      0.070146      0.282511      2.844904   
std        2.227997      5.376434      0.340861      0.858168      6.187177   
min        0.000000      0.000000      0.000000      0.000000      0.000000   
25%        0.000000      0.000000      0.000000      0.000000      0.000000   
50%        0.000000      0.000000      0.000000      0.000000      1.000000   
75%        1.000000      1.000000      0.000000      0.000000      3.000000   
max       41.000000    235.000000     10.000000     29.000000    235.000000   

       Unsecured_TL      Other_TL  Age_Oldest_TL  Age_Newest_TL  
count  51336.000000  51336.000000   51336.000000   51336.000000  
mean       2.013694      1.089762     -32.575639     -62.149525  
std        3.198322      2.417496    2791.869609    2790.818622  
min        0.000000      0.000000  -99999.000000  -99999.000000  
25%        0.000000      0.000000      14.000000       4.000000  
50%        1.000000      0.000000      33.000000       8.000000  
75%        2.000000      1.000000      64.000000      17.000000  
max       55.000000     80.000000     392.000000     392.000000  

[8 rows x 26 columns]

So, Age_Oldest_TL and Age_Newest_TL have some values as -99999.
lets see how many rows are -99999 in these two columns

In [ ]:
count = len(df1[df1['Age_Oldest_TL'] == -99999])    # copied command
count


40

Out of 51336, only 40 are missing values(-99999) so i will just remove those rows

In [34]:
# Remove nulls
df1 = df1.loc[df1['Age_Oldest_TL'] != -99999]

In [35]:
df1.describe()

PROSPECTID      Total_TL  Tot_Closed_TL  Tot_Active_TL  \
count  51296.000000  51296.000000   51296.000000   51296.000000   
mean   25668.574353      4.861471       2.772321       2.089149   
std    14819.611579      7.179141       5.943536       2.291375   
min        1.000000      1.000000       0.000000       0.000000   
25%    12836.750000      1.000000       0.000000       1.000000   
50%    25666.500000      2.000000       1.000000       1.000000   
75%    38503.250000      5.000000       3.000000       3.000000   
max    51336.000000    235.000000     216.000000      47.000000   

       Total_TL_opened_L6M  Tot_TL_closed_L6M  pct_tl_open_L6M  \
count         51296.000000       51296.000000     51296.000000   
mean              0.737328           0.429254         0.184718   
std               1.297059           0.990286         0.297486   
min               0.000000           0.000000         0.000000   
25%               0.000000           0.000000         0.000000   
50%               0.000000           0.000000         0.000000   
75%               1.000000           1.000000         0.312000   
max              27.000000          19.000000         1.000000   

       pct_tl_closed_L6M  pct_active_tl  pct_closed_tl  ...         CC_TL  \
count       51296.000000   51296.000000   51296.000000  ...  51296.000000   
mean            0.089164       0.577349       0.422651  ...      0.125078   
std             0.205700       0.379804       0.379804  ...      0.505386   
min             0.000000       0.000000       0.000000  ...      0.000000   
25%             0.000000       0.250000       0.000000  ...      0.000000   
50%             0.000000       0.556000       0.444000  ...      0.000000   
75%             0.054000       1.000000       0.750000  ...      0.000000   
max             1.000000       1.000000       1.000000  ...     27.000000   

        Consumer_TL       Gold_TL       Home_TL         PL_TL    Secured_TL  \
count  51296.000000  51296.000000  51296.000000  51296.000000  51296.000000   
mean       1.136970      1.563046      0.070181      0.282712      2.846752   
std        2.228639      5.378357      0.340963      0.858461      6.189165   
min        0.000000      0.000000      0.000000      0.000000      0.000000   
25%        0.000000      0.000000      0.000000      0.000000      0.000000   
50%        0.000000      0.000000      0.000000      0.000000      1.000000   
75%        1.000000      1.000000      0.000000      0.000000      3.000000   
max       41.000000    235.000000     10.000000     29.000000    235.000000   

       Unsecured_TL      Other_TL  Age_Oldest_TL  Age_Newest_TL  
count  51296.000000  51296.000000   51296.000000   51296.000000  
mean       2.014718      1.089773      45.376969      15.780022  
std        3.199333      2.418314      41.737151      21.995676  
min        0.000000      0.000000       0.000000       0.000000  
25%        0.000000      0.000000      14.000000       4.000000  
50%        1.000000      0.000000      33.000000       8.000000  
75%        2.000000      1.000000      64.000000      17.000000  
max       55.000000     80.000000     392.000000     392.000000  

[8 rows x 26 columns]

Now even the Age_Newest_TL column has no -99999 values.
That means all those values were against each other

In [36]:
df2.describe()

PROSPECTID  time_since_recent_payment  time_since_first_deliquency  \
count  51336.000000               51336.000000                 51336.000000   
mean   25668.500000               -8129.961314                -70020.091320   
std    14819.571046               27749.328514                 45823.312757   
min        1.000000              -99999.000000                -99999.000000   
25%    12834.750000                  46.000000                -99999.000000   
50%    25668.500000                  70.000000                -99999.000000   
75%    38502.250000                 161.000000                     8.000000   
max    51336.000000                6065.000000                    35.000000   

       time_since_recent_deliquency  num_times_delinquent  \
count                  51336.000000          51336.000000   
mean                  -70022.375838              1.573749   
std                    45819.820741              4.165012   
min                   -99999.000000              0.000000   
25%                   -99999.000000              0.000000   
50%                   -99999.000000              0.000000   
75%                        3.000000              1.000000   
max                       35.000000             74.000000   

       max_delinquency_level  max_recent_level_of_deliq  num_deliq_6mts  \
count           51336.000000               51336.000000    51336.000000   
mean           -70003.987085                  13.521953        0.184977   
std             45847.976100                  53.336976        0.710240   
min            -99999.000000                   0.000000        0.000000   
25%            -99999.000000                   0.000000        0.000000   
50%            -99999.000000                   0.000000        0.000000   
75%                15.000000                  10.000000        0.000000   
max               900.000000                 900.000000       12.000000   

       num_deliq_12mts  num_deliq_6_12mts  ...  PL_utilization       PL_Flag  \
count     51336.000000       51336.000000  ...    51336.000000  51336.000000   
mean          0.480053           0.295076  ...   -86556.225194      0.167874   
std           1.522210           1.027471  ...    34111.414750      0.373758   
min           0.000000           0.000000  ...   -99999.000000      0.000000   
25%           0.000000           0.000000  ...   -99999.000000      0.000000   
50%           0.000000           0.000000  ...   -99999.000000      0.000000   
75%           0.000000           0.000000  ...   -99999.000000      0.000000   
max          28.000000          20.000000  ...        1.708000      1.000000   

       pct_PL_enq_L6m_of_L12m  pct_CC_enq_L6m_of_L12m  pct_PL_enq_L6m_of_ever  \
count            51336.000000            51336.000000            51336.000000   
mean                 0.190414                0.065182                0.170492   
std                  0.376218                0.235706                0.350209   
min                  0.000000                0.000000                0.000000   
25%                  0.000000                0.000000                0.000000   
50%                  0.000000                0.000000                0.000000   
75%                  0.000000                0.000000                0.000000   
max                  1.000000                1.000000                1.000000   

       pct_CC_enq_L6m_of_ever  max_unsec_exposure_inPct       HL_Flag  \
count            51336.000000              51336.000000  51336.000000   
mean                 0.056302             -45127.943635      0.271116   
std                  0.213506              49795.784556      0.444540   
min                  0.000000             -99999.000000      0.000000   
25%                  0.000000             -99999.000000      0.000000   
50%                  0.000000                  0.333000      0.000000   
75%                  0.000000                  2.164250      1.000000   
max                  1.000000       

many columns have missing values
"time_since_recent_payment","time_since_first_deliquency","time_since_recent_deliquency","max_delinquency_level","PL_utilization","max_unsec_exposure_inPct" etc, lets list them all

In [42]:
for col in df2.columns:
    count = (df2[col] == -99999).sum()
    if count > 0:
        print(f"{col} has {count} values equal to -99999")



time_since_recent_payment has 4291 values equal to -99999
time_since_first_deliquency has 35949 values equal to -99999
time_since_recent_deliquency has 35949 values equal to -99999
max_delinquency_level has 35949 values equal to -99999
max_deliq_6mts has 12890 values equal to -99999
max_deliq_12mts has 10832 values equal to -99999
tot_enq has 6321 values equal to -99999
CC_enq has 6321 values equal to -99999
CC_enq_L6m has 6321 values equal to -99999
CC_enq_L12m has 6321 values equal to -99999
PL_enq has 6321 values equal to -99999
PL_enq_L6m has 6321 values equal to -99999
PL_enq_L12m has 6321 values equal to -99999
time_since_recent_enq has 6321 values equal to -99999
enq_L12m has 6321 values equal to -99999
enq_L6m has 6321 values equal to -99999
enq_L3m has 6321 values equal to -99999
pct_currentBal_all_TL has 72 values equal to -99999
CC_utilization has 47636 values equal to -99999
PL_utilization has 44435 values equal to -99999
max_unsec_exposure_inPct has 23178 values equal to -

many columns has 10k+ , 30k+, 40k+ missing values. our dataset has total 51k rows. 
removing all columns with more than 10k missing values

In [43]:
columns_to_be_removed = []

for i in df2.columns:
    if df2.loc[df2[i] == -99999].shape[0] > 10000:
        columns_to_be_removed .append(i)

df2 = df2.drop(columns_to_be_removed, axis =1)

still missing values must be there, so we will just remove those rows.

In [45]:
for i in df2.columns:
    df2 = df2.loc[ df2[i] != -99999 ]

Now we dont have any missing values in both data frames

# Merging these two data frames

In [47]:
for i in list(df1.columns):
    if i in list(df2.columns):
        print (i)

PROSPECTID


In [48]:
# Merge the two dataframes, inner join so that no nulls are present
df = pd. merge ( df1, df2, how ='inner', left_on = ['PROSPECTID'], right_on = ['PROSPECTID'] )

In [49]:
df.shape

(42064, 79)

In [50]:
# check how many columns are categorical
for i in df.columns:
    if df[i].dtype == 'object':
        print(i)

MARITALSTATUS
EDUCATION
GENDER
last_prod_enq2
first_prod_enq2
Approved_Flag


In [54]:
# Chi-square test
for i in ['MARITALSTATUS', 'EDUCATION', 'GENDER', 'last_prod_enq2', 'first_prod_enq2']:
    chi2, pval, _, _ = chi2_contingency(pd.crosstab(df[i], df['Approved_Flag']))
    print(i, '---', pval)


MARITALSTATUS --- 3.578180861038862e-233
EDUCATION --- 2.6942265249737532e-30
GENDER --- 1.907936100186563e-05
last_prod_enq2 --- 0.0
first_prod_enq2 --- 7.84997610555419e-287


Since all the categorical features have pval <=0.05, we will accept all

In [57]:
# VIF for numerical columns
numeric_columns = []
for i in df.columns:
    if df[i].dtype != 'object' and i not in ['PROSPECTID','Approved_Flag']:
        numeric_columns.append(i)


In [59]:
# VIF sequentially check

vif_data = df[numeric_columns]
total_columns = vif_data.shape[1]
columns_to_be_kept = []
column_index = 0

for i in range (0,total_columns):
    
    vif_value = variance_inflation_factor(vif_data, column_index)
    print (column_index,'---',vif_value)
    
    
    if vif_value <= 6:
        columns_to_be_kept.append( numeric_columns[i] )
        column_index = column_index+1
    
    else:
        vif_data = vif_data.drop([ numeric_columns[i] ] , axis=1)

c:\Github\Credit Risk Prediction\.venv\Lib\site-packages\statsmodels\stats\outliers_influence.py:197: RuntimeWarning: divide by zero encountered in scalar divide
  vif = 1. / (1. - r_squared_i)


0 --- inf


c:\Github\Credit Risk Prediction\.venv\Lib\site-packages\statsmodels\stats\outliers_influence.py:197: RuntimeWarning: divide by zero encountered in scalar divide
  vif = 1. / (1. - r_squared_i)


0 --- inf
0 --- 11.320180023967996
0 --- 8.363698035000336
0 --- 6.520647877790928
0 --- 5.149501618212625
1 --- 2.611111040579735


c:\Github\Credit Risk Prediction\.venv\Lib\site-packages\statsmodels\stats\outliers_influence.py:197: RuntimeWarning: divide by zero encountered in scalar divide
  vif = 1. / (1. - r_squared_i)


2 --- inf
2 --- 1788.7926256209232
2 --- 8.601028256477228
2 --- 3.8328007921530785
3 --- 6.0996533816467355
3 --- 5.5813520096427585
4 --- 1.9855843530987785


c:\Github\Credit Risk Prediction\.venv\Lib\site-packages\statsmodels\stats\outliers_influence.py:197: RuntimeWarning: divide by zero encountered in scalar divide
  vif = 1. / (1. - r_squared_i)


5 --- inf
5 --- 4.809538302819343
6 --- 23.270628983464636
6 --- 30.595522588100053
6 --- 4.384346405965583
7 --- 3.0646584155234238
8 --- 2.898639771299253
9 --- 4.377876915347324
10 --- 2.2078535836958433
11 --- 4.916914200506864
12 --- 5.214702030064725
13 --- 3.3861625024231476
14 --- 7.840583309478997
14 --- 5.255034641721438


c:\Github\Credit Risk Prediction\.venv\Lib\site-packages\statsmodels\stats\outliers_influence.py:197: RuntimeWarning: divide by zero encountered in scalar divide
  vif = 1. / (1. - r_squared_i)


15 --- inf
15 --- 7.380634506427232
15 --- 1.4210050015175733
16 --- 8.083255010190323
16 --- 1.6241227524040114
17 --- 7.257811920140003
17 --- 15.59624383268298
17 --- 1.8258570471324314
18 --- 1.5080839450032664
19 --- 2.172088834824578
20 --- 2.623397553527229
21 --- 2.2959970812106167
22 --- 7.360578319196446
22 --- 2.1602387773102563
23 --- 2.8686288267891467
24 --- 6.458218003637277
24 --- 2.8474118865638265
25 --- 4.753198156284083
26 --- 16.22735475594825
26 --- 6.424377256363877
26 --- 8.887080381808687
26 --- 2.3804746142952653
27 --- 8.60951347651454
27 --- 13.06755093547673
27 --- 3.5000400566546555
28 --- 1.9087955874813773
29 --- 17.006562234161628
29 --- 10.730485153719197
29 --- 2.3538497522950266
30 --- 22.104855915136433
30 --- 2.7971639638512906
31 --- 3.424171203217696
32 --- 10.175021454450935
32 --- 6.408710354561301
32 --- 1.001151196262562
33 --- 3.069197305397274
34 --- 2.8091261600643724
35 --- 20.249538381980678
35 --- 15.864576541593774
35 --- 1.83316497405

In [60]:
# check Anova for columns_to_be_kept 

from scipy.stats import f_oneway

columns_to_be_kept_numerical = []

for i in columns_to_be_kept:
    a = list(df[i])  
    b = list(df['Approved_Flag'])  
    
    group_P1 = [value for value, group in zip(a, b) if group == 'P1']
    group_P2 = [value for value, group in zip(a, b) if group == 'P2']
    group_P3 = [value for value, group in zip(a, b) if group == 'P3']
    group_P4 = [value for value, group in zip(a, b) if group == 'P4']


    f_statistic, p_value = f_oneway(group_P1, group_P2, group_P3, group_P4)

    if p_value <= 0.05:
        columns_to_be_kept_numerical.append(i)


feature selection is done for cat and num features

In [63]:
# listing all the final features
features = columns_to_be_kept_numerical + ['MARITALSTATUS', 'EDUCATION', 'GENDER', 'last_prod_enq2', 'first_prod_enq2']
df = df[features + ['Approved_Flag']]
df.sample(5)

pct_tl_open_L6M  pct_tl_closed_L6M  Tot_TL_closed_L12M  \
19397              0.0              0.000                   0   
31954              1.0              0.000                   0   
31186              0.0              0.333                   1   
32679              0.0              0.400                   2   
5194               0.0              0.000                   0   

       pct_tl_closed_L12M  Tot_Missed_Pmnt  CC_TL  Home_TL  PL_TL  Secured_TL  \
19397               0.000                0      0        0      0           0   
31954               0.000                0      0        0      0           1   
31186               0.333                0      0        0      0           2   
32679               0.400                0      0        0      0           2   
5194                0.000                0      0        0      0           2   

       Unsecured_TL  ...  pct_PL_enq_L6m_of_ever  pct_CC_enq_L6m_of_ever  \
19397             1  ...                    0.00                     0.0   
31954             0  ...                    0.00                     0.0   
31186             1  ...                    0.25                     0.0   
32679             3  ...                    0.00                     0.0   
5194              0  ...                    0.00                     0.0   

       HL_Flag  GL_Flag  MARITALSTATUS       EDUCATION  GENDER  \
19397        0        0        Married            12TH       M   
31954        0        0        Married          OTHERS       M   
31186        1        0        Married  UNDER GRADUATE       M   
32679        0        0        Married   POST-GRADUATE       F   
5194         1        0        Married    PROFESSIONAL       M   

       last_prod_enq2  first_prod_enq2  Approved_Flag  
19397              PL               PL             P2  
31954    ConsumerLoan     ConsumerLoan             P3  
31186              PL               AL             P2  
32679    ConsumerLoan     ConsumerLoan             P2  
5194     ConsumerLoan     ConsumerLoan             P2  

[5 rows x 43 columns]

In [66]:
df.shape

(42064, 43)

# Label encoding for the categorical features
['MARITALSTATUS', 'EDUCATION', 'GENDER' , 'last_prod_enq2' ,'first_prod_enq2']

In [68]:

df['MARITALSTATUS'].unique()

array(['Married', 'Single'], dtype=object)

In [69]:
df['EDUCATION'].unique()

array(['12TH', 'GRADUATE', 'SSC', 'POST-GRADUATE', 'UNDER GRADUATE',
       'OTHERS', 'PROFESSIONAL'], dtype=object)

In [70]:
df['GENDER'].unique()

array(['M', 'F'], dtype=object)

In [71]:
df['last_prod_enq2'].unique()

array(['PL', 'ConsumerLoan', 'AL', 'CC', 'others', 'HL'], dtype=object)

In [72]:
df['first_prod_enq2'].unique()

array(['PL', 'ConsumerLoan', 'others', 'AL', 'HL', 'CC'], dtype=object)

Ordinal feature -- EDUCATION
SSC            : 1
12TH           : 2
GRADUATE       : 3
UNDER GRADUATE : 3
POST-GRADUATE  : 4
OTHERS         : 1
PROFESSIONAL   : 3

Others has to be verified by the business end user 

In [73]:
df.loc[df['EDUCATION'] == 'SSC',['EDUCATION']]              = 1
df.loc[df['EDUCATION'] == '12TH',['EDUCATION']]             = 2
df.loc[df['EDUCATION'] == 'GRADUATE',['EDUCATION']]         = 3
df.loc[df['EDUCATION'] == 'UNDER GRADUATE',['EDUCATION']]   = 3
df.loc[df['EDUCATION'] == 'POST-GRADUATE',['EDUCATION']]    = 4
df.loc[df['EDUCATION'] == 'OTHERS',['EDUCATION']]           = 1
df.loc[df['EDUCATION'] == 'PROFESSIONAL',['EDUCATION']]     = 3

In [ ]:
df['EDUCATION'] = df['EDUCATION'].astype(int)   #Converting column type to int

In [75]:
df_encoded = pd.get_dummies(df, columns=['MARITALSTATUS','GENDER', 'last_prod_enq2' ,'first_prod_enq2'])

In [81]:
df_encoded.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 42064 entries, 0 to 42063
Data columns (total 55 columns):
 #   Column                        Non-Null Count  Dtype  
---  ------                        --------------  -----  
 0   pct_tl_open_L6M               42064 non-null  float64
 1   pct_tl_closed_L6M             42064 non-null  float64
 2   Tot_TL_closed_L12M            42064 non-null  int64  
 3   pct_tl_closed_L12M            42064 non-null  float64
 4   Tot_Missed_Pmnt               42064 non-null  int64  
 5   CC_TL                         42064 non-null  int64  
 6   Home_TL                       42064 non-null  int64  
 7   PL_TL                         42064 non-null  int64  
 8   Secured_TL                    42064 non-null  int64  
 9   Unsecured_TL                  42064 non-null  int64  
 10  Other_TL                      42064 non-null  int64  
 11  Age_Oldest_TL                 42064 non-null  int64  
 12  Age_Newest_TL                 42064 non-null  int64  
 13  t

In [80]:
df_encoded.shape

(42064, 55)

In [82]:
df_encoded.to_csv("C:/Github/Credit Risk Prediction/data/processed/df_encoded.csv", index=False)